# Лабораторная работа №3
#### Выполнил: Нистратов Д.Г. гр.9383 
#### Вариант 12
## Цель работы 
Практическое применение библиотеки символьных вычислений Python (SymPy) для решения задач дискретной математики и математической логики.
## Задачи лабораторной работы
1. Решить свой вариант диофантова уравнения.
2. Имея открытую часть ключа RSA найти соответствующую закрытую часть и расшифровать кодированное слово.
3. Записать в формате ДНФ и КНФ булеву функцию, заданную своим аналитическим выражением.

##  Выполнение работы

### Диофантово уравнение
Решить диофантово уравнение: $$ 1819x+ 1734y = −85 $$

In [6]:
import sympy
from sympy import symbols
from sympy.solvers.diophantine import diophantine
x, y = symbols("x, y", integer=True)
diophantine(1819*x + 1734*y + 85)

{(102*t_0 - 205, 215 - 107*t_0)}

Ответ: $$ x = 102t - 205$$ $$ y = 215 - 107t$$

### Имея открытую часть ключа RSA найти соответствующую закрытую часть и расшифровать кодированное слово.

Русское слово из четырех букв закодировано при помощи алгоритма RSA открытым ключом (e = 7,m = 39). Шифрованное сообщение имеет вид (13; 19; 32; 18). Подберите закрытую часть ключа и прочитайте исходное слово. Буквам русского алфавита соответсвуют числа в диапазоне от 2 до 32 (исключены буквы “e” и “ъ”).

In [17]:
from sympy.crypto.crypto import rsa_public_key
from sympy.crypto.crypto import decipher_rsa

message = [13, 19, 32, 18]
key = rsa_public_key(39, 7)

abc = ("а б в г д e ж з и й к л м н о п р с т у ф х ц ч ш щ ы ь ъ э ю я").split()

msg = []

msg.append([abc[decipher_rsa(number, key) - 2] for number in message])

msg

[['л', 'e', 'т', 'у']]

Ответ: $$ лету $$

### Записать в формате ДНФ и КНФ булеву функцию, заданную своим аналитическим выражением.

Пусть $$ f(x, y, z) = ((x \newcommand*\xor{\mathbin{\oplus}} y)(z \newcommand*\xor{\mathbin{\oplus}} x)) ∨ ((z ∨ y)y): $$



In [21]:
from sympy.logic.boolalg import to_dnf, to_cnf

x, y, z = symbols('x y z')
expression = ((x ^ y) & (z ^ x)) | ((z | y) & y) 

In [22]:
to_dnf(expression, simplify=True)

y | (x & ~z)

In [23]:
to_cnf(expression, simplify=True)

(x | y) & (y | ~z)

### Вывод
В ходе выполнения лабораторной работы было изучено практическое применение библиотеки символьных вычислений Python (SymPy) для решения задач дискретной математики и математической логики.